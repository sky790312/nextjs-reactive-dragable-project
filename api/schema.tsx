export interface PAGE {
  id: number
  name: string
  url: string
}

export interface BLOCK {
  id: number
  name: string
  x: number
  y: number
  o: number
  c: string
}
