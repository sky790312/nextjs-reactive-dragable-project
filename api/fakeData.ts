/**
 * pages should come from somewhere and get all url config.
 */
export const FAKE_PAGES = [
  {
    id: 1,
    name: 'page1',
    url: '/pages/1',
  },
  {
    id: 2,
    name: 'page2',
    url: '/pages/2',
  },
]

/**
 * items should come from api(maybe by page) and get x, y, o style.
 */
export const FAKE_BLOCKS = [
  {
    id: 1,
    name: 'Element 1-1',
    x: 10,
    y: 10,
    o: 1,
    c: '#008800',
  },
  {
    id: 2,
    name: 'Element 1-2',
    x: 60,
    y: 60,
    o: 0.5,
    c: '#008800',
  },
  {
    id: 3,
    name: 'Element 1-3',
    x: 110,
    y: 110,
    o: 1,
    c: '#008800',
  },
  {
    id: 4,
    name: 'Element 2-1',
    x: 30,
    y: 30,
    o: 1,
    c: '#008800',
  },
]
