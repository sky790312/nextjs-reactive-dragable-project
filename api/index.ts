import { FAKE_PAGES, FAKE_BLOCKS } from '@/api/fakeData'
import { PAGE, BLOCK } from '@/api/schema'

export const fetchPages = async () => {
  const data: PAGE[] = await FAKE_PAGES
  return data
}

export const fetchBlocks = async () => {
  const data: BLOCK[] = await FAKE_BLOCKS
  return data
}
