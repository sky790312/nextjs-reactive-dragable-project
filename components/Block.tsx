import React, { useRef } from 'react'
import styled from 'styled-components'
import { useDrag } from '@/hooks/useDrag'
import { BLOCK } from '@/api/schema'
import { Dispatch } from 'redux'
import { blockActions } from '@/store/state.block'
import { useSelector, shallowEqual, useDispatch } from 'react-redux'
import { RootState } from '@/store/index'

interface Props {
  block: BLOCK
}

const Block: React.FC<Props> = React.memo(({ block }) => {
  const dispatch: Dispatch = useDispatch()
  const { setCurrentBlock } = blockActions
  const currentBlock = useSelector(
    (state: RootState) => state.block.currentBlock,
    shallowEqual,
  )
  const isCurrentBlock = block.id === currentBlock?.id
  const startingPosition = { x: block.x, y: block.y }

  const dragRef = useRef(null)
  const {
    isDragging,
    itemPosition,
    handleMouseDown,
    handleMouseMove,
    handleMouseUp,
  } = useDrag(startingPosition, dragRef)

  const onMouseDown = (e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    dispatch(setCurrentBlock(block))
    handleMouseDown(e)
  }

  const onMouseUp = () => {
    const x = Number(itemPosition.left.slice(0, itemPosition.left.length - 2))
    const y = Number(itemPosition.top.slice(0, itemPosition.top.length - 2))
    const updateBlock = {
      ...block,
      x,
      y,
    }
    dispatch(setCurrentBlock(updateBlock))
    handleMouseUp()
  }

  const getBlockPosition = () => {
    let top, left
    if (isCurrentBlock) {
      if (isDragging) {
        top = itemPosition.top
        left = itemPosition.left
      } else {
        top = `${currentBlock.y}px`
        left = `${currentBlock.x}px`
      }
    } else {
      top = `${block.y}px`
      left = `${block.x}px`
    }
    return { top, left }
  }

  return (
    <BlockWrapper
      style={{ ...getBlockPosition() }}
      o={block.o}
      c={block.c}
      isActive={isCurrentBlock}
      ref={dragRef}
      onMouseDown={onMouseDown}
      onMouseMove={handleMouseMove}
      onMouseLeave={handleMouseMove}
      onMouseUp={onMouseUp}
    />
  )
})

export default Block

const BlockWrapper = styled.div<{ o: number; c: string; isActive: boolean }>`
  position: absolute;
  width: 50px;
  height: 50px;
  opacity: ${props => props.o};
  background: ${props => props.c};
  outline: ${props => (props.isActive ? 1 : 0)}px solid
    ${props => props.theme.colors.blue};
`
