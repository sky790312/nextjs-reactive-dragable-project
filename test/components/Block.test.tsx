import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import Block from '@/components/Block'
import { FAKE_BLOCKS } from '@/api/fakeData'

describe('Block component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      withProviders(<Block block={FAKE_BLOCKS[0]} />),
      {},
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
