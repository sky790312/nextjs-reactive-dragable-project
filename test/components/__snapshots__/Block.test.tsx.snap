// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Block__BlockWrapper-sc-816v2v-0 ofJXj"
    style="top: 10px; left: 10px;"
  />
</DocumentFragment>
`;
